src/iptvsimple/utilities/StreamUtils.cpp

```diff
@@ -63,7 +63,41 @@ void StreamUtils::SetAllStreamProperties(PVR_NAMED_VALUE* properties, unsigned i
     }
     else // inputstream.adpative
     {
-      StreamUtils::SetStreamProperty(properties, propertiesCount, propertiesMax, PVR_STREAM_PROPERTY_STREAMURL, streamURL);
+      // If no media headers are explicitly set for inputstream.adaptive,
+      // strip the headers from streamURL and put it to media headers property
+
+      if (channel.GetProperty("inputstream.adaptive.stream_headers").empty()) {
+        // preserve any kodi protocol options after "|"
+        std::string url = streamURL;
+        std::string protocolOptions;
+        std::string untrimmed_header = "";
+
+        size_t found = streamURL.find_first_of('|');
+        if (found != std::string::npos) {
+          url = streamURL.substr(0, found);
+          protocolOptions = streamURL.substr(found + 1, streamURL.length());
+
+          std::vector<std::string> headers = StringUtils::Split(protocolOptions, "&");
```

Comment:
This is very confusing. Are you stripping off and splitting each header only to add them all back the same way? If so why?

If you did need to add headers again there is a function for that already but I don't think it's needed at all. 
https://github.com/kodi-pvr/pvr.iptvsimple/blob/48ace1a75a72ff83e7baea7d23ea90bfc2334df6/src/iptvsimple/utilities/StreamUtils.cpp#L215-L240

Reply:
I am splitting the header and url-encoding the value to match https://github.com/peak3d/inputstream.adaptive/blob/Matrix/src/helpers.cpp#L459

Makes sense to upgrade the function you mentioned with "urlencode" boolean parameter, though.

Reply:
Ya, you should upgrade it 😉

You could also add another function to StreamUtils to encode the protocol options and that would keep this function clean and allow the url encoring be used by other inputstreams in the future.

Reply:
I added new function and tested the build.